### ISSUE TITLE
Inlining fails when there is no default branch

### ISSUE THREAD

**Issue by DianQK** [llvm/llvm-project#76772]
From https://github.com/llvm/llvm-project/pull/76669#issuecomment-1873266792.

If a version with a default branch can be inlined, it should also get inlined without a default branch. Generally, a switch without a default branch generates fewer instructions.

I haven't created an example yet. I'm using this issue to track this regression.

cc @dtcxzyw

**Comment by DianQK** [comment]
I checked the relevant code and I found a lot of interesting things.

#### 1. `InlineCost.cpp` does not consider the default branch.

https://github.com/llvm/llvm-project/blob/7954c57124b495fbdc73674d71f2e366e4afe522/llvm/lib/Analysis/InlineCost.cpp#L2459-L2464

https://github.com/llvm/llvm-project/blob/7954c57124b495fbdc73674d71f2e366e4afe522/llvm/include/llvm/CodeGen/BasicTTIImpl.h#L448-L457

Godbolt: https://llvm.godbolt.org/z/fd17b4nMj

**Comment by dtcxzyw** [comment]
> I checked the relevant code and I found a lot of interesting things.
> 
> #### 1. `InlineCost.cpp` does not consider the default branch.
> https://github.com/llvm/llvm-project/blob/7954c57124b495fbdc73674d71f2e366e4afe522/llvm/lib/Analysis/InlineCost.cpp#L2459-L2464
> 
> https://github.com/llvm/llvm-project/blob/7954c57124b495fbdc73674d71f2e366e4afe522/llvm/include/llvm/CodeGen/BasicTTIImpl.h#L448-L457
> 
> Godbolt: https://llvm.godbolt.org/z/fd17b4nMj

Yeah, we should add an extra cost for the reachable default case. The inline cost of switch should be lower than the original form after #76669.

**Comment by DianQK** [comment]
> > I checked the relevant code and I found a lot of interesting things.
> > #### 1. `InlineCost.cpp` does not consider the default branch.
> Yeah, we should add an extra cost for the reachable default case. The inline cost of switch should be lower than the original form after #76669.

If both will become lookup tables or neither will become lookup tables. I think so. This test case happens to encounter a boundary value. The cost of a normal switch is much less than a lookup table. Sad...

Anyway, I'll start by creating two draft PRs for the first two issues to illustrate my ideas.

**Comment by DianQK** [comment]
> ~~If both will become lookup tables or neither will become lookup tables. I think so. This test case happens to encounter a boundary value. The cost of a normal switch is much less than a lookup table.~~

But the unreachable default branch should always be no worse than the default reachable branch. So this must be the result of inconsistent information between the inline and lookup tables.

**Comment by DianQK** [comment]
Let's move on to step two.

**Comment by DianQK** [comment]
> 1. InlineCost.cpp does not consider the default branch.
> 2. Inconsistent minimum cases for lookup table
> 3. Large cost of instructions for lookup tables?

I think I've figured out the problem here. Among the three issues I mentioned earlier, the second and third ones are maybe incorrect. The thresholds/costs for these two problems are related to the backend and are unrelated to the intermediate IR. For example, the lookup table here isn't transformed into a GEP instruction in the IR; instead, it's a machine code jump instruction: https://llvm.godbolt.org/z/Ps3Mhxjcx.

So, this result seems expected as well. I'm going to close this issue. Unless we encounter actual performance regressions, we can reconsider how to add additional rules for handling. (Perhaps we should revisit the third problem, aiming to reduce the cost overhead of the backend lookup table. *Or converting the small switch statement with no default branch into several comparison instructions.*)

cc @dtcxzyw
